Data/Aeson/Types/Class.hs

```diff
@@ -261,13 +276,50 @@ data ToJSONKeyFunction a
     = ToJSONKeyText (a -> Text, a -> Encoding)
     | ToJSONKeyValue (a -> Value, a -> Encoding)
 
+-- | With GHC 7.8 + we carry around 'Coercible Text a' dictionary,
+-- to have even some amount of safety net.
+-- Unfortunately we cannot enforce that 'Hashable' instance agree on the type level
+--
+-- ATM this type is intentionally not exported. FromJSONKeyFunction can be inspected,
+-- but cannot be constructed.
+data CoerceText a where
+#if HAS_COERCIBLE
+    CoerceText :: Coercible Text a => CoerceText a
+#else
+    CoerceText :: CoerceText a
+#endif
+
 data FromJSONKeyFunction a
-    = FromJSONKeyCoerce
+    = FromJSONKeyCoerce (CoerceText a)
     | FromJSONKeyText (Text -> a)
     | FromJSONKeyTextParser (Text -> Parser a)
     | FromJSONKeyValue (Value -> Parser a)
-  deriving (Functor)
 
+instance Functor FromJSONKeyFunction where
```

Comment:
Not a valid `Functor` instance since the following does not hold:

    fmap id x = x